### ISSUE TITLE
🛠️ Tooling: Refactor bingo-systems to unlock repository hosts other than GitHub

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [bingo-js/bingo#77]
### Bug Report Checklist

- [X] I have pulled the latest `main` branch of the repository.
- [X] I have [searched for related issues](https://github.com/JoshuaKGoldberg/create/issues?q=is%3Aissue) and found none that matched my issue.

### Overview

Right now, Bingo requires the user be authenticated with GitHub to create a repository. An [Octokit](https://octokit.github.io/rest.js) instance is created during runtime always. Which is reasonable and convenient for feature parity with create-typescript-app and the single place I personally always put my repositories on, GitHub...

But there are other authentication and repository hosts out there! At least:

* Azure DevOps
* GitLab
* Theoretically others that people will want to ad hoc plug in on their own

Long-term, the Bingo engine will need to be able to support that. I don't know yet what that will look like.

### Additional Info

It's not going to happen in [milestone/2 Blocks Launch](https://github.com/JoshuaKGoldberg/create/milestone/2), and I'm not sure whether it'll even happen in [milestone/3 Broad Launch](https://github.com/JoshuaKGoldberg/create/milestone/3). Marking as blocked for a while. But it definitely should happen.

This came in part out of talking with @johnnyreilly about the work he's been doing on his personal and corporate templates. https://johnnyreilly.com/npx-and-azure-artifacts-the-secret-cli-delivery-mechanism talks a good amount about Azure DevOps / Azure Artifacts hooks.

Co-authored-by: @johnnyreilly

**Comment by JoshuaKGoldberg** [comment]
@allcontributors add @johnnyreilly for ideas

**Comment by johnnyreilly** [comment]
I'll be well placed to help with the Azure DevOps side of things when the time comes - I now know quite a lot about that world!

**Comment by JoshuaKGoldberg** [comment]
Ok! Now that requests are serializable and well-typed (#65 -> #278), repositories describe their API calls as plain old objects with `type: "fetch" | "octokit"` per https://www.create.bingo/build/packages/bingo-requests. Nothing's stopping us from adding in a new type, say `"ado"` or `"gitlab"`, with its own well-typed requests. 🚀 Issue unblocked!

**Comment by JoshuaKGoldberg** [comment]
Thinking on this: right now, the only hosting platform natively supported by Bingo templates is GitHub:

* [`bingo-requests`](https://www.create.bingo/build/packages/bingo-requests) only has one well-typed request kind, [`CreatedOctokitRequest`](https://www.create.bingo/build/packages/bingo-requests#createdoctokitrequest)
* [`bingo-systems`](https://www.create.bingo/build/packages/bingo-systems#createsystemfetchers) has one general system fetchers authentication property, [`auth`](https://www.create.bingo/build/packages/bingo-systems#createsystemfetchers-auth), used for GitHub

I think in order to support hosts such as Azure DevOps and GitLab, we'll need to make `bingo-systems` more generalized the way `bingo-requests` is. As in, instead of having one general `auth` property, we'll want to make the property specific to GitHub. That'll allow adding in other properties for ADO, GL, etc.

This is not a trivial refactor. It requires deep knowledge of Bingo's systems. If you're reading this and thinking _"I don't get it"_ then you'll probably want to tackle other, not-so-deep issues first. But technically accepting PRs for any enterprising individual to try out. 🚀

**Comment by JoshuaKGoldberg** [comment]
💡 Another note: I think long-term it might be best to completely separate auth/hosts out of Bingo altogether. The list of hosts folks might want is virtually infinite: ADO, GL, GH, some-closed-sourced-company-thing, ... I think a lean Bingo core with a clean, rich extensibility story is best.

Just thinking out loud, maybe long-term something like a `@bingo/github` / `bingo-github` should exist? It might contain:
* Logic for how to retrieve the running user's GitHub token
* Hooks for Bingo itself to default known options with GitHub: e.g. for `owner` to default to the currently running user
* Hooks for `prepareOptions` functions to with GitHub: e.g. for a `description` option to source from a GitHub API call for a repo's description
* Type-safe constructions for requests to be sent _if_ GitHub auth is available

I don't yet know how that API would look. I suspect it'd be best done as a followup issue after this one. Just noting here so the info isn't lost & can be thought of when designing for this issue.
